**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант №12

Набор данных: movies.csv

Атрибуты:
...

### 1.Чтение файла (набора данных)

In [2]:
# импорт библиотек, чтение файла с помощью pandas
import pandas as pd

data = pd.read_csv('movies.csv', sep=';')


### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

In [3]:
# применить метод head
data.head(20)

title  release_year  score  \
0      David Attenborough: A Life on Our Planet        2020.0    9.0   
1                                     Inception        2010.0    8.8   
2                                  Forrest Gump        1994.0    8.8   
3                                    Anbe Sivam        2003.0    8.7   
4                            Bo Burnham: Inside        2021.0    8.7   
5                           Saving Private Ryan        1998.0    8.6   
6                              Django Unchained        2012.0    8.4   
7                                        Dangal        2016.0    8.4   
8                        Bo Burnham: Make Happy        2016.0    8.4   
9                         Louis C.K.: Hilarious        2010.0    8.4   
10              Dave Chappelle: Sticks & Stones        2019.0    8.4   
11                                     3 Idiots        2009.0    8.4   
12                                 Black Friday        2004.0    8.4   
13                                 Super Deluxe        2019.0    8.4   
14  Winter on Fire: Ukraine's Fight for Freedom        2015.0    8.3   
15                  Once Upon a Time in America        1984.0    8.3   
16                                  Taxi Driver        1976.0    8.3   
17                          Like Stars on Earth        2007.0    8.3   
18                            Bo Burnham: What.        2013.0    8.3   
19                            Full Metal Jacket        1987.0    8.3   

    number_of_votes  duration   main_genre main_production  
0           31180.0        83  documentary              GB  
1         2268288.0       148        scifi              GB  
2         1994599.0       142        drama              US  
3           20595.0       160       comedy              IN  
4           44074.0        87       comedy              US  
5         1346020.0       169        drama              US  
6         1472668.0       165      western              US  
7          180247.0       161       action              IN  
8           14356.0        60       comedy              US  
9           11973.0        84       comedy              US  
10          25687.0        65       comedy              US  
11         385782.0       170       comedy              IN  
12          20611.0       143        crime              IN  
13          13680.0       176     thriller              IN  
14          17710.0        98  documentary              UA  
15         342335.0       229        drama              US  
16         795222.0       113        crime              US  
17         188234.0       165        drama              IN  
18          11488.0        60       comedy              US  
19         723306.0       116        drama              GB

2.2 Оценка данных с помощью метода info.

In [4]:
# выполнит метод info
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 389 entries, 0 to 388
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   title            384 non-null    object 
 1   release_year     387 non-null    float64
 2   score            386 non-null    float64
 3   number_of_votes  388 non-null    float64
 4   duration         389 non-null    int64  
 5   main_genre       389 non-null    object 
 6   main_production  388 non-null    object 
dtypes: float64(3), int64(1), object(3)
memory usage: 21.4+ KB


2.3 Оценка данных с помощью метода describe.

In [5]:
# оцените числовые столбцы с помощью describe

data.describe()

release_year       score  number_of_votes    duration
count    387.000000  386.000000     3.880000e+02  389.000000
mean    2010.976744    7.512176     1.391523e+05  123.352185
std       10.805726    0.443870     2.365279e+05   28.304800
min     1954.000000    6.900000     1.013900e+04   28.000000
25%     2007.500000    7.100000     2.055375e+04  104.000000
50%     2014.000000    7.400000     4.606950e+04  122.000000
75%     2018.000000    7.800000     1.551585e+05  139.000000
max     2022.000000    9.000000     2.268288e+06  229.000000


---

**Сделать выводы. Вы должны понимать, что означает тот или иной столбец, чтобы
 ответить на вопросы на защите.**


 ---


Выводы по обзору данных (пп. 1-4):

Набор данных содержит информацию о 389 фильмах с 7 атрибутами:

title: Название фильма (категориальный/строковый). Присутствуют пропуски (384 non-null).
release_year: Год выпуска фильма (числовой, тип float64). Присутствуют пропуски (387 non-null).
score: Оценка фильма (числовой, тип float64). Присутствуют пропуски (386 non-null).
number_of_votes: Количество голосов (числовой, тип float64). Присутствуют пропуски (388 non-null).
duration: Продолжительность фильма в минутах (числовой, тип int64). Пропусков нет.
main_genre: Основной жанр фильма (категориальный/строковый). Пропусков нет.
main_production: Основная страна производства фильма (категориальный/строковый). Присутствуют пропуски (388 non-null).
Типы данных:

Числовые: release_year, score, number_of_votes, duration.
Категориальные/Строковые: title, main_genre, main_production.
Пропуски обнаружены в столбцах title, release_year, score, number_of_votes и main_production. Количество пропусков незначительно по сравнению с общим количеством записей (389).

 2.4 Оценка названий столбцов

In [6]:
# Вывести на экран названия столбцов с помощью df.columns. Выявить проблемы с названиями, если они есть. При необходимости переименовать столбцы. Если проблемы не обнаружены также дать пояснения.

data.columns

Index(['title', 'release_year', 'score', 'number_of_votes', 'duration',
       'main_genre', 'main_production'],
      dtype='object')

In [7]:
# переименование при необходимости

### 3. Проверка пропусков

In [8]:
# Проверить данные на наличие пропусков и устранить их, если они есть (пропуски необходимо либо удалить, либо заменить каким-то значением).
print(data.isnull().sum())

title              5
release_year       2
score              3
number_of_votes    1
duration           0
main_genre         0
main_production    1
dtype: int64


In [9]:
data.dropna(inplace=True)

print("\nКоличество пропусков в каждом столбце после удаления:")
print(data.isnull().sum())


Количество пропусков в каждом столбце после удаления:
title              0
release_year       0
score              0
number_of_votes    0
duration           0
main_genre         0
main_production    0
dtype: int64



---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с пропусками, а также знать другие способы работы с пропусками, чтобы
 ответить на вопросы на защите.**


 ---

Были удалены все записи с пропущенными значениями, тк в основном они были категориальные, а также их было не так много, соответственно на анализ они  не повлиют.

Также можно было заполнить пропуски медианным значением или модой у категориальных признаков.


### 4. Проверка дубликатов

#### Проверка явных дубликатов

In [10]:
data.duplicated().sum()

np.int64(2)

In [11]:
# удалите дубликаты, если они есть
data.drop_duplicates(inplace=True)

# Проверим еще раз количество явных дубликатов после удаления
print("Количество явных дубликатов после удаления:", data.duplicated().sum())

Количество явных дубликатов после удаления: 0


#### Проверка неявных дубликатов

In [12]:
# Проверка на неявные дубликаты в категориальных столбцах
print("Уникальные значения в столбце 'title':")
print(data['title'].unique()[:20]) # Выводим первые 20 уникальных значений для обзора

print("\nУникальные значения в столбце 'main_genre':")
print(data['main_genre'].unique())

print("\nУникальные значения в столбце 'main_production':")
print(data['main_production'].unique())

Уникальные значения в столбце 'title':
['David Attenborough: A Life on Our Planet' 'Inception' 'Forrest Gump'
 'Anbe Sivam' 'Bo Burnham: Inside' 'Saving Private Ryan'
 'Django Unchained' 'Dangal' 'Bo Burnham: Make Happy'
 'Louis C.K.: Hilarious' 'Dave Chappelle: Sticks & Stones' '3 Idiots'
 'Black Friday' 'Super Deluxe'
 "Winter on Fire: Ukraine's Fight for Freedom"
 'Once Upon a Time in America' 'Taxi Driver' 'Like Stars on Earth'
 'Bo Burnham: What.' 'Full Metal Jacket']

Уникальные значения в столбце 'main_genre':
['documentary' 'scifi' 'drama' 'comedy' 'western' 'action' 'crime'
 'thriller' 'war' 'fantasy' 'romance' 'horror' 'animation' 'sports'
 'COMEDY' 'dramaa' 'musical']

Уникальные значения в столбце 'main_production':
['GB' 'US' 'IN' 'UA' 'CD' 'TR' 'ES' 'AU' 'JP' 'ZA' 'HK' 'DE' 'KR' 'CA'
 'BE' 'NO' 'NZ' 'MX' 'FR' 'MW' 'TH' 'AR' 'PS' 'HU' 'IT' 'CN' 'PL' 'KH'
 'IE' 'BR' 'XX' 'LT' 'NL' 'DK' 'ID']


In [13]:
# удалите дубликаты, если они есть
data['main_genre'] = data['main_genre'].replace({'COMEDY': 'comedy', 'dramaa': 'drama'})

# Проверим уникальные значения еще раз
print("\nУникальные значения в столбце 'main_genre' после исправления:")
print(data['main_genre'].unique())


Уникальные значения в столбце 'main_genre' после исправления:
['documentary' 'scifi' 'drama' 'comedy' 'western' 'action' 'crime'
 'thriller' 'war' 'fantasy' 'romance' 'horror' 'animation' 'sports'
 'musical']


---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с дубликатами, знать, что такое явные и неявные дубликаты и способы работы с ними, чтобы ответить на вопросы на защите.**


 ---

Явные дубликаты — это строки, которые полностью повторяются в наборе данных. Удалил их с помощью метода drop_duplicates(), потому что они представляют собой идентичные записи. Неявные дубликаты — это записи, которые по сути одинаковы, но имеют небольшие отличия (например, "Comedy" и "COMEDY"). Обнаружил такие неявные дубликаты в столбце main_genre ('COMEDY' и 'dramaa' вместо 'drama').

Чтобы устранить неявные дубликаты, можно привести все строковые значения к единому регистру (например, нижнему), удалить лишние пробелы или использовать метод replace() для замены некорректных значений на правильные ('COMEDY' на 'comedy', 'dramaa' на 'drama').


### 5. Провека типов данных

In [14]:
data

title  release_year  score  \
0    David Attenborough: A Life on Our Planet        2020.0    9.0   
1                                   Inception        2010.0    8.8   
2                                Forrest Gump        1994.0    8.8   
3                                  Anbe Sivam        2003.0    8.7   
4                          Bo Burnham: Inside        2021.0    8.7   
..                                        ...           ...    ...   
382                                  The Trip        2021.0    6.9   
383                                  The Dirt        2019.0    6.9   
384                                   Top Gun        1986.0    6.9   
385                               Radhe Shyam        2022.0    6.9   
386                       Sorry to Bother You        2018.0    6.9   

     number_of_votes  duration   main_genre main_production  
0            31180.0        83  documentary              GB  
1          2268288.0       148        scifi              GB  
2          1994599.0       142        drama              US  
3            20595.0       160       comedy              IN  
4            44074.0        87       comedy              US  
..               ...       ...          ...             ...  
382          19706.0       113       comedy              NO  
383          47603.0       108        drama              US  
384         329656.0       110        drama              US  
385          21328.0       138      romance              IN  
386          75653.0       111      fantasy              US  

[381 rows x 7 columns]

In [15]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 381 entries, 0 to 386
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   title            381 non-null    object 
 1   release_year     381 non-null    float64
 2   score            381 non-null    float64
 3   number_of_votes  381 non-null    float64
 4   duration         381 non-null    int64  
 5   main_genre       381 non-null    object 
 6   main_production  381 non-null    object 
dtypes: float64(3), int64(1), object(3)
memory usage: 23.8+ KB


In [16]:
data['release_year'] = data['release_year'].astype(int)
data['number_of_votes'] = data['number_of_votes'].astype(int)

print(data.info())

<class 'pandas.core.frame.DataFrame'>
Index: 381 entries, 0 to 386
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   title            381 non-null    object 
 1   release_year     381 non-null    int64  
 2   score            381 non-null    float64
 3   number_of_votes  381 non-null    int64  
 4   duration         381 non-null    int64  
 5   main_genre       381 non-null    object 
 6   main_production  381 non-null    object 
dtypes: float64(1), int64(3), object(3)
memory usage: 23.8+ KB
None


In [17]:
# Проверьте типы данных, при необходимости измените типы данных, чтобы они соответствовали действительности.

---

**Обратите внимание, что во всех вариантах необходимо сделать приведение типов. Будьте готовы на защите аргументировать проверку типов (почему выполнены те или иные преобразования).**


 ---

Обоснование выбора типов: release_year, number_of_votes - целочисленные значения

### 6. Группировка данных

#### Задание 1

Группировка - год выхода и количество фильмов каждого жанра
(main_genre).

In [18]:
genre_counts_by_year = data.groupby(['release_year', 'main_genre'])['score'].mean().reset_index()
print("Количество фильмов по году выхода и жанру:")
print(genre_counts_by_year)


Количество фильмов по году выхода и жанру:
     release_year main_genre  score
0            1954    romance   7.50
1            1961        war   7.50
2            1964      drama   7.80
3            1966    western   7.30
4            1967      drama   7.70
..            ...        ...    ...
159          2021   thriller   7.55
160          2022     comedy   7.30
161          2022      crime   7.20
162          2022      drama   7.50
163          2022    romance   6.90

[164 rows x 3 columns]


In [19]:
# выполните группировку согласно варианту

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

На основе результатов группировки по году выхода и жанру (genre_counts_by_year) можно увидеть распределение фильмов по жанрам в разные годы. Например, можно определить:

 - Какие жанры были популярны в определенные годы.
 - Как менялось количество фильмов определенного жанра с течением времени.
 - В какие годы выходило больше всего фильмов в целом.


---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 2

*`Формулировка задания`*

Группировка - год выхода и количество фильмов каждого жанра
(main_genre). Создать датафрейм. Переименовать столбец с количеством в
“сount”. Отсортировать по убыванию столбца “count”.

In [20]:
genre_counts_by_year = data.groupby(['release_year', 'main_genre']).size().reset_index(name='count')

genre_counts_by_year_sorted = genre_counts_by_year.sort_values(by='count', ascending=False)

print("Количество фильмов по году выхода и жанру (отсортировано по количеству):")
print(genre_counts_by_year_sorted)

Количество фильмов по году выхода и жанру (отсортировано по количеству):
     release_year main_genre  count
121          2017      drama     13
139          2019      drama     13
147          2020      drama      9
103          2015      drama      9
155          2021      drama      9
..            ...        ...    ...
157          2021      scifi      1
158          2021     sports      1
160          2022     comedy      1
161          2022      crime      1
163          2022    romance      1

[164 rows x 3 columns]


In [21]:
# выполните группировку согласно варианту

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

Результат этой группировки (genre_counts_by_year_sorted), отсортированный по убыванию количества фильмов, позволяет быстро увидеть наиболее "продуктивные" сочетания года выпуска и жанра. Можно определить:

 - В какие годы и в каких жанрах было выпущено наибольшее количество фильмов.
 - Наиболее популярные жанры в определенные периоды времени с точки зрения объема производства.





---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 3

*`Формулировка задания`*

Сводная таблица (pivot_table) - средний рейтинг (score) фильмов по
жанрам. Отсортировать по убыванию рейтинга. Округлить до двух знаков.

In [22]:
genre_score_pivot = data.pivot_table(index='main_genre', values='score', aggfunc='mean')

genre_score_pivot_sorted = genre_score_pivot.sort_values(by='score', ascending=False)

genre_score_pivot_sorted['score'] = genre_score_pivot_sorted['score'].round(2)

print("Средний рейтинг фильмов по жанрам (отсортировано по убыванию рейтинга):")
print(genre_score_pivot_sorted)

Средний рейтинг фильмов по жанрам (отсортировано по убыванию рейтинга):
             score
main_genre        
war           7.80
scifi         7.77
documentary   7.76
action        7.70
comedy        7.61
western       7.56
fantasy       7.48
drama         7.48
crime         7.47
thriller      7.47
romance       7.40
sports        7.40
horror        7.35
animation     7.20
musical       7.00


In [23]:
# выполните сводную таблицу согласно варианту

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной сводной таблицы?`**

На основе результатов сводной таблицы (genre_score_pivot_sorted) можно увидеть средний рейтинг фильмов для каждого жанра, отсортированный от самого высокого к самому низкому. Это позволяет определить:

 - Какие жанры имеют наивысший средний рейтинг среди зрителей (например, "war" и "scifi" в данном случае).
 - Какие жанры имеют более низкий средний рейтинг (например, "animation" и "musical").
 - Сравнить средние рейтинги разных жанров между собой.




---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 4

*`Формулировка задания`*

Сводная таблица (pivot_table) - средний рейтинг (score) фильмов по
жанрам - строки и годам - столбцы. Отсортировать по возрастанию main_genre.


In [24]:
genre_year_score_pivot = data.pivot_table(index='main_genre', columns='release_year', values='score', aggfunc='mean')

genre_year_score_pivot_sorted = genre_year_score_pivot.sort_index(ascending=True)

print("Средний рейтинг фильмов по жанрам (строки) и годам (столбцы):")
print(genre_year_score_pivot_sorted)

Средний рейтинг фильмов по жанрам (строки) и годам (столбцы):
release_year  1954  1961  1964  1966  1967  1971  1973  1975  1976  1979  ...  \
main_genre                                                                ...   
action         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
animation      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
comedy         NaN   NaN   NaN   NaN   NaN   NaN   NaN   8.2   NaN   7.8  ...   
crime          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   8.3   NaN  ...   
documentary    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
drama          NaN   NaN   7.8   NaN   7.7   NaN   NaN   NaN   NaN   NaN  ...   
fantasy        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
horror         NaN   NaN   NaN   NaN   NaN   NaN   8.1   NaN   NaN   NaN  ...   
musical        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
romance        7.5   NaN   NaN   NaN   NaN   Na

In [25]:
# выполните сводную таблицу согласно варианту

**`Сделайте вывод - интерпретацию - что можно увидеть из результата выполнения данного кода?`**

На основе результатов выполнения кода, который создает сводную таблицу средних рейтингов фильмов по жанрам (строки) и годам выпуска (столбцы), можно увидеть:

 - как менялся средний рейтинг определенного жанра из года в год.
 - какие жанры имели более высокие или низкие средние рейтинги в определенный год.
 - Годы с наибольшей активностью по жанрам


---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

### Вывод


***`Сделать расширенный вывод. `***

Вывод должен описывать результаты исследования: какой набор данных вы анализировали (предметная область), какие методы предобработки применили, какие выводы из группировок и сводных таблиц можно сделать (они могут повторять выводы выше) и т.п. Объём вывода, как правило не более 0.5-0.75 страницы А4, всё таки вывод должен быть конкретным (без теории и “воды”).


В рамках лабораторной работы был проанализирован набор данных movies.csv, содержащий сведения о 389 фильмах: название, год выпуска, рейтинг, количество голосов, продолжительность, жанр и страну производства.

Предобработка данных включала загрузку набора в DataFrame, анализ структуры и удаление строк с пропусками и дубликатами. Были устранены неявные дубликаты в категориальных признаках (например, некорректные значения жанров), приведены типы данных к корректным (например, release_year и number_of_votes — к целочисленным). После очистки в датасете осталось 381 запись без пропусков и дубликатов.

Группировки и сводные таблицы позволили выявить ключевые закономерности:

 - Количество фильмов по годам и жанрам показало периоды наибольшей активности кинопроизводства и доминирующие жанры в разные годы.

 - Отсортированная группировка подтвердила, что в 2017–2019 годах жанр drama был наиболее представлен.

 - Сводная таблица среднего рейтинга по жанрам показала, что наивысшие оценки получили жанры war, scifi и documentary, тогда как animation и musical имели более низкие значения.

 - Анализ динамики рейтингов по жанрам и годам позволил проследить изменения зрительских предпочтений.

В результате была проведена полная предобработка и базовый анализ данных о фильмах. Полученные сводки и группировки дают представление о распределении жанров, динамике их популярности и различиях в оценках.

### Дополнительное задание

1. Создать столбец “Категория года выхода (release_year)” (с помощью
категоризации). Выделить минимум 3 категории (свежий, старый, средний),
фильтрацию для уровня года выбрать самостоятельно, аргументировать
выбор. Создать сводную таблицу: средний и медианный год выхода по
жанру и категории года выхода.


In [26]:
def categorize_year(year):
    if year < 2000:
        return 'старый'
    elif 2000 <= year < 2015:
        return 'средний'
    else:
        return 'свежий'

data['Категория года выхода (release_year)'] = data['release_year'].apply(categorize_year)

pivot_table_year_category = data.pivot_table(
    index=['main_genre', 'Категория года выхода (release_year)'],
    values='release_year',
    aggfunc=['mean', 'median']
)

print("Сводная таблица: средний и медианный год выхода по жанру и категории года выхода")
print(pivot_table_year_category)

Сводная таблица: средний и медианный год выхода по жанру и категории года выхода
                                                         mean       median
                                                 release_year release_year
main_genre  Категория года выхода (release_year)                          
action      свежий                                2017.333333       2016.0
            средний                               2011.000000       2011.0
            старый                                1998.000000       1998.0
animation   свежий                                2018.500000       2018.5
            средний                               2009.000000       2009.0
comedy      свежий                                2018.200000       2018.0
            средний                               2009.041667       2009.0
            старый                                1985.375000       1984.5
crime       свежий                                2018.800000       2019.5
            средний

**Вывод по дополнительному заданию:**

Был создан новый столбец "Категория года выхода (release_year)" путем категоризации года выпуска фильмов на "старый" (до 2000 года), "средний" (с 2000 по 2014 год включительно) и "свежий" (с 2015 года). Выбранные границы категорий основаны на распределении годов выпуска в исходном наборе данных.

Затем была построена сводная таблица, отображающая средний и медианный год выпуска фильмов по каждому жанру и созданной категории года выхода. Результаты этой таблицы позволяют увидеть:

- Можно наблюдать, какие жанры имеют больше "старых", "средних" или "свежих" фильмов. Например, жанры с большим количеством "свежих" фильмов могут указывать на их текущую популярность или активность производства.
- Разница между средним и медианным годом выпуска в каждой группе может указывать на наличие выбросов или неравномерное распределение годов выпуска в данной категории жанра.
- Анализируя средние и медианные значения по категориям года выхода для каждого жанра, можно получить представление о том, как менялся "возраст" фильмов в различных жанрах.


4. Создать столбец “Категория года выхода (release_year)” (с помощью
категоризации). Выделить минимум 3 категории (свежий, старый, средний),
фильтрацию для уровня года выбрать самостоятельно, аргументировать
выбор. Создать группировку: средний, минимальный, максимальный и
медианный score по категории года выхода.


In [27]:
def categorize_year(year):
    if year < 2000:
        return 'старый'
    elif 2000 <= year < 2015:
        return 'средний'
    else:
        return 'свежий'

score_by_year_category = data.groupby('Категория года выхода (release_year)')['score'].agg(['mean', 'min', 'max', 'median'])

print("Статистика рейтинга (score) по категории года выхода:")
print(score_by_year_category)

Статистика рейтинга (score) по категории года выхода:
                                          mean  min  max  median
Категория года выхода (release_year)                            
свежий                                7.430526  6.9  9.0     7.3
средний                               7.580272  6.9  8.8     7.5
старый                                7.611364  6.9  8.8     7.6


**Вывод по дополнительному заданию:**

Был создан новый столбец "Категория года выхода (release_year)" путем категоризации года выпуска фильмов на "старый" (до 2000 года), "средний" (с 2000 по 2014 год включительно) и "свежий" (с 2015 года). Выбранные границы категорий основаны на распределении годов выпуска в исходном наборе данных, чтобы обеспечить достаточное количество фильмов в каждой категории для анализа.

Затем была создана группировка данных по созданной категории года выхода, и для каждой категории были рассчитаны средний, минимальный, максимальный и медианный рейтинги (score). Результаты этой группировки позволяют увидеть:

- Можно сравнить средние, минимальные, максимальные и медианные рейтинги фильмов в каждой категории ("старый", "средний", "свежий"). Это позволяет понять, есть ли зависимость между "возрастом" фильма и его рейтингом.
- Минимальное и максимальное значения показывают разброс рейтингов в каждой группе, что помогает оценить однородность оценок.
- Среднее и медианное значения дают представление о типичном рейтинге фильмов в каждой категории.

9. Создать столбец “Категория длительности фильма” (с помощью
категоризации). Выделить минимум 3 категории (короткий, длинный,
средний), фильтрацию для уровня длительности выбрать самостоятельно,
аргументировать выбор. Создать группировку: средний, минимальный,
максимальный и медианный score по категории длительности .


In [28]:
def categorize_duration(duration):
    if duration < 90:
        return 'короткий'
    elif 90 <= duration <= 120:
        return 'средний'
    else:
        return 'длинный'

data['Категория длительности фильма'] = data['duration'].apply(categorize_duration)

score_by_duration_category = data.groupby('Категория длительности фильма')['score'].agg(['mean', 'min', 'max', 'median'])

print("Статистика рейтинга (score) по категории длительности фильма:")
print(score_by_duration_category)

Статистика рейтинга (score) по категории длительности фильма:
                                   mean  min  max  median
Категория длительности фильма                            
длинный                        7.572222  6.9  8.8    7.50
короткий                       7.800000  7.0  9.0    7.85
средний                        7.364901  6.9  8.3    7.30



Был создан новый столбец "Категория длительности фильма" путем категоризации длительности фильмов на "короткий" (менее 90 минут), "средний" (от 90 до 120 минут включительно) и "длинный" (более 120 минут). Выбранные границы категорий основаны на общем распределении длительности фильмов в наборе данных.

Затем была создана группировка данных по созданной категории длительности фильма, и для каждой категории были рассчитаны средний, минимальный, максимальный и медианный рейтинги (score). Результаты этой группировки позволяют увидеть:

- Можно сравнить средние, минимальные, максимальные и медианные рейтинги фильмов в каждой категории ("короткий", "средний", "длинный"). Это позволяет понять, есть ли зависимость между длительностью фильма и его рейтингом.
- Минимальное и максимальное значения показывают разброс рейтингов в каждой группе, что помогает оценить однородность оценок в каждой категории длительности.
- Среднее и медианное значения дают представление о типичном рейтинге фильмов в каждой категории длительности.

На основе полученных данных можно сделать выводы о том, как средний, минимальный, максимальный и медианный рейтинги фильмов различаются в зависимости от их категории длительности.

11. Создать столбец “Категория длительности фильма” (с помощью
категоризации). Выделить минимум 3 категории (короткий, длинный,
средний), фильтрацию для уровня длительности выбрать самостоятельно,
аргументировать выбор. Создать сводную таблицу: средний, минимальный
и медианный number_of_votes по категории длительности и жанру

In [29]:
def categorize_duration(duration):
    if duration < 90:
        return 'короткий'
    elif 90 <= duration <= 120:
        return 'средний'
    else:
        return 'длинный'

if 'Категория длительности фильма' not in data.columns:
    data['Категория длительности фильма'] = data['duration'].apply(categorize_duration)

pivot_table_votes_duration_genre = data.pivot_table(
    index=['Категория длительности фильма', 'main_genre'],
    values='number_of_votes',
    aggfunc=['mean', 'min', 'median']
)

print("Сводная таблица: среднее, минимальное и медианное количество голосов по категории длительности и жанру")
print(pivot_table_votes_duration_genre)

Сводная таблица: среднее, минимальное и медианное количество голосов по категории длительности и жанру
                                                     mean             min  \
                                          number_of_votes number_of_votes   
Категория длительности фильма main_genre                                    
длинный                       action         85755.333333           24681   
                              comedy         57536.000000           10528   
                              crime         165260.583333           10139   
                              documentary    31638.500000           14605   
                              drama         155800.435644           10656   
                              fantasy       108942.000000           17716   
                              horror        262059.333333           32814   
                              romance        32356.000000           10206   
                              scifi         731471


Был создан новый столбец "Категория длительности фильма" путем категоризации длительности фильмов на "короткий" (менее 90 минут), "средний" (от 90 до 120 минут включительно) и "длинный" (более 120 минут). Выбранные границы категорий основаны на общем распределении длительности фильмов в наборе данных.

Затем была создана сводная таблица, отображающая среднее, минимальное и медианное количество голосов (number_of_votes) по каждой категории длительности фильма и жанру. Результаты этой сводной таблицы позволяют увидеть:

- Можно наблюдать, как среднее, минимальное и медианное количество голосов различается для фильмов разных жанров в пределах каждой категории длительности.
- Сравнивая значения в сводной таблице, можно определить, какие жанры в каждой категории длительности имеют в среднем больше или меньше голосов.
- Анализируя данные, можно предположить, есть ли связь между длительностью фильма и количеством голосов, которые он получает, в зависимости от его жанра.


15.  Создать столбец “Категория рейтинга” (score) (с помощью категоризации).
Выделить минимум 3 категории (высокий, низкий, средний), фильтрацию
для уровня рейтинга выбрать самостоятельно, аргументировать выбор.
Создать сводную таблицу: средний, минимальный и медианный release_year
по категории рейтинга и жанру.

In [30]:
def categorize_score(score):
    if score >= 8.0:
        return 'высокий'
    elif 7.0 <= score < 8.0:
        return 'средний'
    else:
        return 'низкий'

data['Категория рейтинга'] = data['score'].apply(categorize_score)

pivot_table_year_score_genre = data.pivot_table(
    index=['Категория рейтинга', 'main_genre'],
    values='release_year',
    aggfunc=['mean', 'min', 'median']
)

print("Сводная таблица: средний, минимальный и медианный год выпуска по категории рейтинга и жанру")
print(pivot_table_year_score_genre)

Сводная таблица: средний, минимальный и медианный год выпуска по категории рейтинга и жанру
                                       mean          min       median
                               release_year release_year release_year
Категория рейтинга main_genre                                        
высокий            action       2018.500000         2016       2018.5
                   comedy       2009.058824         1975       2013.0
                   crime        1999.666667         1976       2004.0
                   documentary  2017.666667         2014       2018.0
                   drama        2006.230769         1984       2009.0
                   fantasy      2013.000000         2010       2012.0
                   horror       1973.000000         1973       1973.0
                   romance      2005.500000         1995       2005.5
                   scifi        2007.000000         1997       2007.0
                   thriller     2013.363636         2006       2015.

Был создан новый столбец "Категория рейтинга" путем категоризации рейтинга фильмов на "высокий" (рейтинг 8.0 и выше), "средний" (от 7.0 до 7.9 включительно) и "низкий" (ниже 7.0). Выбранные границы категорий основаны на распределении рейтингов в исходном наборе данных.

Затем была создана сводная таблица, отображающая средний, минимальный и медианный год выпуска фильмов по каждой категории рейтинга и жанру. Результаты этой сводной таблицы позволяют увидеть:

- Можно наблюдать, как средний, минимальный и медианный год выпуска различается для фильмов разных жанров в пределах каждой категории рейтинга.
- Сравнивая значения в сводной таблице, можно определить, какие жанры в каждой категории рейтинга имеют в среднем более старые или более новые фильмы.
- Анализируя данные, можно предположить, есть ли связь между рейтингом фильма, его жанром и годом выпуска. Например, являются ли высокорейтинговые фильмы определенных жанров, как правило, более новыми или более старыми.


18. Создать столбец “Категория рейтинга” (score) (с помощью категоризации).
Выделить минимум 3 категории (высокий, низкий, средний), фильтрацию
для уровня рейтинга выбрать самостоятельно, аргументировать выбор.
Выполнить фильтрацию: выбрать только записи с высокий и средний
рейтинг + топ 5 жанров с наибольшей популярностью (количество записей)
+ год выхода в определенном диапазоне (диапазон подобрать
самостоятельно, чтобы было достаточное количество записей) . Создать
сводные таблицы: средний, минимальный, максимальный и медианный
рейтинг по категории рейтинга и жанру.


In [31]:
def categorize_score(score):
    if score >= 8.0:
        return 'высокий'
    elif 7.0 <= score < 8.0:
        return 'средний'
    else:
        return 'низкий'

if 'Категория рейтинга' not in data.columns:
    data['Категория рейтинга'] = data['score'].apply(categorize_score)

top_genres = data['main_genre'].value_counts().nlargest(5).index.tolist()
print(f"Топ 5 жанров по популярности: {top_genres}")

start_year = 2000
end_year = 2020
print(f"Выбранный диапазон годов выпуска: с {start_year} по {end_year}")

# Выполняем фильтрацию данных
filtered_data = data[
    (data['Категория рейтинга'].isin(['высокий', 'средний'])) &
    (data['main_genre'].isin(top_genres)) &
    (data['release_year'] >= start_year) &
    (data['release_year'] <= end_year)
]

display(filtered_data.head())


pivot_table_filtered = filtered_data.pivot_table(
    index=['Категория рейтинга', 'main_genre'],
    values='score',
    aggfunc=['mean', 'min', 'max', 'median']
)

print(pivot_table_filtered)

Топ 5 жанров по популярности: ['drama', 'thriller', 'comedy', 'crime', 'documentary']
Выбранный диапазон годов выпуска: с 2000 по 2020


title  release_year  score  \
0   David Attenborough: A Life on Our Planet          2020    9.0   
3                                 Anbe Sivam          2003    8.7   
8                     Bo Burnham: Make Happy          2016    8.4   
9                      Louis C.K.: Hilarious          2010    8.4   
10           Dave Chappelle: Sticks & Stones          2019    8.4   

    number_of_votes  duration   main_genre main_production  \
0             31180        83  documentary              GB   
3             20595       160       comedy              IN   
8             14356        60       comedy              US   
9             11973        84       comedy              US   
10            25687        65       comedy              US   

   Категория года выхода (release_year) Категория длительности фильма  \
0                                свежий                      короткий   
3                               средний                       длинный   
8                                свежий                      короткий   
9                               средний                      короткий   
10                               свежий                      короткий   

   Категория рейтинга  
0             высокий  
3             высокий  
8             высокий  
9             высокий  
10            высокий

                                    mean   min   max median
                                   score score score  score
Категория рейтинга main_genre                              
высокий            comedy       8.253846   8.0   8.7   8.20
                   crime        8.300000   8.2   8.4   8.30
                   documentary  8.360000   8.1   9.0   8.20
                   drama        8.100000   8.0   8.3   8.10
                   thriller     8.090909   8.0   8.4   8.00
средний            comedy       7.388889   7.0   7.9   7.30
                   crime        7.392857   7.0   7.8   7.35
                   documentary  7.500000   7.1   7.9   7.40
                   drama        7.397753   7.0   7.9   7.40
                   thriller     7.337838   7.0   7.8   7.30


Был создан новый столбец "Категория рейтинга" путем категоризации рейтинга фильмов на "высокий" (рейтинг 8.0 и выше), "средний" (от 7.0 до 7.9 включительно) и "низкий" (ниже 7.0).

Затем была выполнена фильтрация данных для выбора записей с "высоким" и "средним" рейтингом, принадлежащих к топ-5 жанрам по популярности (Drama, Thriller, Comedy, Crime, Documentary), и выпущенных в диапазоне с 2000 по 2020 год включительно.

На основе отфильтрованных данных была создана сводная таблица, отображающая средний, минимальный, максимальный и медианный рейтинг по категории рейтинга и жанру. Результаты этой сводной таблицы позволяют увидеть:

- Можно сравнить средние, минимальные, максимальные и медианные рейтинги фильмов в категориях "высокий" и "средний" для топ-5 жанров в указанном временном диапазоне.
- Минимальное и максимальное значения показывают диапазон рейтингов в каждой комбинации категории рейтинга и жанра после фильтрации.
- Среднее и медианное значения дают представление о типичном рейтинге фильмов в каждой группе.


22. Создать столбец “Категория длительности фильма” (с помощью
категоризации). Выделить минимум 3 категории (короткий, длинный,
средний), фильтрацию для уровня длительности выбрать самостоятельно,
аргументировать выбор. Выполнить фильтрацию: выбрать только записи
рейтингом от 5 до 8, а также длинные и средние фильмы + топ 7 жанров по
среднему рейтингу . Создать сводные таблицы: средний, минимальный,
максимальный и медианный score по категории длительности и жанру и
средний, минимальный, максимальный и медианный number_of_votes по
категории длительности и жанру.

In [32]:
def categorize_duration(duration):
    if duration < 90:
        return 'короткий'
    elif 90 <= duration <= 120:
        return 'средний'
    else:
        return 'длинный'

if 'Категория длительности фильма' not in data.columns:
    data['Категория длительности фильма'] = data['duration'].apply(categorize_duration)

top_genres_by_score = data.groupby('main_genre')['score'].mean().nlargest(7).index.tolist()
print(f"Топ 7 жанров по среднему рейтингу: {top_genres_by_score}")

filtered_data_2 = data[
    (data['score'] >= 5.0) & (data['score'] <= 8.0) &
    (data['Категория длительности фильма'].isin(['длинный', 'средний'])) &
    (data['main_genre'].isin(top_genres_by_score))
]

display(filtered_data_2.head())

pivot_table_score_duration_genre = filtered_data_2.pivot_table(
    index=['Категория длительности фильма', 'main_genre'],
    values='score',
    aggfunc=['mean', 'min', 'max', 'median']
)

print("\nСводная таблица: статистика рейтинга по категории длительности и жанру (отфильтрованные данные)")
print(pivot_table_score_duration_genre)

pivot_table_votes_duration_genre_2 = filtered_data_2.pivot_table(
    index=['Категория длительности фильма', 'main_genre'],
    values='number_of_votes',
    aggfunc=['mean', 'min', 'max', 'median']
)

print("\nСводная таблица: статистика количества голосов по категории длительности и жанру (отфильтрованные данные)")
print(pivot_table_votes_duration_genre_2)

Топ 7 жанров по среднему рейтингу: ['war', 'scifi', 'documentary', 'action', 'comedy', 'western', 'fantasy']


title  release_year  score  number_of_votes  duration  \
59           G.O.R.A.          2004    8.0            61797       127   
61          Vizontele          2001    8.0            36291       110   
72      Life of Brian          1979    8.0           392419        94   
75  Blade Runner 2049          2017    8.0           539864       164   
77         Jab We Met          2007    7.9            51945       138   

   main_genre main_production Категория года выхода (release_year)  \
59      scifi              TR                              средний   
61     comedy              TR                              средний   
72     comedy              GB                               старый   
75      scifi              CA                               свежий   
77     comedy              IN                              средний   

   Категория длительности фильма Категория рейтинга  
59                       длинный            высокий  
61                       средний            высокий  
72                       средний            высокий  
75                       длинный            высокий  
77                       длинный            средний


Сводная таблица: статистика рейтинга по категории длительности и жанру (отфильтрованные данные)
                                               mean   min   max median
                                              score score score  score
Категория длительности фильма main_genre                              
длинный                       action       7.100000   7.1   7.1   7.10
                              comedy       7.400000   6.9   7.9   7.35
                              documentary  7.550000   7.2   7.9   7.55
                              fantasy      7.550000   7.5   7.6   7.55
                              scifi        7.625000   7.2   8.0   7.65
                              war          7.600000   7.5   7.7   7.60
                              western      7.550000   7.3   7.8   7.55
средний                       action       7.100000   7.1   7.1   7.10
                              comedy       7.325000   6.9   8.0   7.25
                              documentary  7.500000

Был создан новый столбец "Категория длительности фильма" путем категоризации длительности фильмов на "короткий" (менее 90 минут), "средний" (от 90 до 120 минут включительно) и "длинный" (более 120 минут).

Затем была выполнена фильтрация данных, чтобы выбрать записи с рейтингом от 5 до 8, а также фильмы категорий "длинный" и "средний", принадлежащие к топ-7 жанрам по среднему рейтингу.

На основе отфильтрованных данных были созданы две сводные таблицы:

1.  **Статистика рейтинга (score) по категории длительности и жанру:** Эта таблица показывает средний, минимальный, максимальный и медианный рейтинг фильмов в отфильтрованных данных для каждой комбинации категории длительности и жанра.
2.  **Статистика количества голосов (number_of_votes) по категории длительности и жанру:** Эта таблица показывает среднее, минимальное, максимальное и медианное количество голосов для фильмов в отфильтрованных данных для каждой комбинации категории длительности и жанра.
